Support for socket.send and intercept callback

noway421 opened the pull request:
That's about it. There's also `__hash__` function added to Peer for convinience and Host.service function is fixed to be in accordance to Host.check_events

aresch commented:
Would it be possible to add tests for the new code?

noway421 commented:
Don't have them on hand, but i'll try to come up with it in several days.

noway421 commented:
Finally got _existing_ tests working, there was a problem with deallocation and the fact that global variable kept the reference and did not allow gc to clear a Host instance. Weak references to the rescue.

noway421 commented:
- added the behaviour of those 2 functions into test_server.py and test_client.py. It makes sense to keep them there because intercept_callback is by design made for something event loop based. Those 2 tests provide loop based approach.
- added 1 unit test for both of them in test_unit.py too
- fixed a couple of bugs along the way
- all is working and it looks good to me. Would be happy to hear out if you have any concerns. 

@aresch

aresch commented on test_enet.py:
```diff
@@ -71,6 +71,24 @@ def test_connect(self):
+
+        while not self.send_done:
+
+            self.client.service(0)
```
You likely want to follow some of the other examples here, maybe the broadcast test below? You should be checking if the event type is connect and then do the socket.send() call, then you can call the server.service() in subsequent loops waiting for the intercept call back to fire.  If you don't add that conditional in, you'll be calling socket.send() over and over again.

noway421 replied:
Hmm, good point, ok.

aresch commented on test_server.py:
```diff
@@ -1,12 +1,18 @@
 host = enet.Host(enet.Address(b"localhost", 54301), 10, 0, 0, 0)
+def receive_callback(address, data):
+    global host
+    if data and data == "\xff\xff\xff\xffgetstatus\x00":
```
Have you tested this with both python2 and 3? Do you need to use a bytestring literal here?

noway421 replied:
Bytestring literal is what we are using in the game server, i.e. i'm copy pasting functionality from there so that tests are closer to production

aresch commented on test_client.py:
```diff
@@ -6,6 +6,21 @@
 host = enet.Host(None, 1, 0, 0, 0)
 peer = host.connect(enet.Address(b"localhost", 54301), 1)
+def receive_callback(address, data):
+    global host
```
This shouldn't be necessary.

noway421 commented on enet.pyx:
```diff
@@ -198,6 +218,13 @@ cdef class Socket:
+        cdef ENetBuffer buffer
+        buffer.data = <void*>(<char*>data)
+        buffer.dataLength = len(data)
+        cdef int result = enet_socket_send(self._enet_socket,
```
Hmm, maybe.

noway421 commented on enet.pyx:
```diff
@@ -798,6 +828,9 @@ cdef class Event:
             return self._packet
 
+from weakref import WeakValueDictionary
+cdef Host_static_instances = WeakValueDictionary()
```
It's a C callback, so not. We need to map ENetHost instances to pyenet Host instances in order to be able to store python data structures and have them available by C _enet_host handle.

aresch replied:
Ok, I see.

Can we change this to host_static_instances?

noway421 commented on enet.pyx:
```diff
@@ -819,6 +852,8 @@ cdef class Host:
 
     cdef ENetHost *_enet_host
     cdef bool dealloc
+    cdef object _interceptCallback
```
ENetHost.intercept is a C function, and in pyenet we want to trigger a python function. We can't simply assign a python function to a variable that expects C function. That's why we store a reference to python function in ENetHost.

noway421 commented on enet.pyx:
```diff
@@ -819,6 +852,8 @@ cdef class Host:
     cdef ENetHost *_enet_host
     cdef bool dealloc
+    cdef object _interceptCallback
+    cdef object __weakref__
```
This is for a weak reference for Host_instances_static variable. I'll try removing it and seeing whether the weak references dictionary will still exists.

We need a weak reference dictionary to map all instances of ENetHost to pyenet Host instances. Could have been done with a normal dictionary, but then there are deallocation problems (I.e. without weak ref dict the Host instances wouldn't be garbage collected)

noway421 replied:
Tested without it, this is actually needed in order for weak references to work. Here's bit of cython documentation back up this claim: http://cython.readthedocs.io/en/latest/src/reference/extension_types.html#weak-referencing

noway421 commented on enet.pyx:
```diff
@@ -1000,6 +1042,27 @@ cdef class Host:
 
+    property intercept:
+        def __get__(self):
+            return self._interceptCallback
```
It's a C function, so unfortunately not. In this specific `property`, we need to operate with python functions because this is what user will be assigning/accessing from a `.py` script.

noway421 commented on test_client.py:
```diff
@@ -20,15 +35,17 @@
 
     counter += 1
     if counter >= MSG_NUMBER:
-        msg = SHUTDOWN_MSG
```
Ok. Wanted to do it after the loop, that's why it's going during last iteration. Will make it in another loop.

noway421 commented on test_enet.py:
```diff
@@ -71,6 +71,24 @@ def test_connect(self):
 
+    def test_socketsend(self):
+
+        self.send_done = False
```
Maybe it can, I'll try.

noway421 replied:
It is needed, because send_done is defined in another scope. The easiest way to access a variable from another scope is to have a reference to it through an object. `self` is a perfect choice for such object.

noway421 commented:
Splitted up tests more, so they are more like new tests, instead of modifying existing ones , and fixed those small things from the feedback.

Now unit test waits for connection in a loop and only then starts interception requests too.

I've tried python 3, but run into problems with weak references. i'll definetely will look into it, but is there a hard requirement of py3k compatibility?

aresch commented:
Yes, Python 3 is a hard requirement.  Python 2 is deprecated and should no longer be used.

noway421 commented:
I'll look into py3k soon-ish, the problem was weak references. At [piqueserver](https://github.com/piqueserver/piqueserver), py3k is WIP too so it's important.

noway421 commented:
For some reason, test_server.py doesn't get intercept_callbacks. py2k does it perfectly, while py3k test_server is silent

dunno why's that happening really...

noway421 commented:
alright, figured it out, it was because `"SHUTDOWN" == b"SHUTDOWN"` was there

Pretty much it's all done. Works both py2k and py3k.

The only thing is that `peer.disconnect()` and peer send packet SHUTDOWN need better treatment (in the tests). Need to think about how to run them from receive_callback so that it sends SHUTDOWN first, then gets acknowledgement, only then disconnects while doing it all in a receive_callback. Just trying to reuse the testing paradigm you are using where loop does `event = host.service(1000)` on each iteration, while incorporating that kinda out-of-the-loop `receive_callback`

Meanwhile you can have a go at another review iteration :-)

noway421 commented:
Sorted out the tests. Made the disconnect logic for intercept_callback test really neat: disconnect is scheduled in the callback and then performed in the main loop.

While looking at it, I found problem that the tests are non-deterministic. First, I thought that tests might be failing sometimes for me because of different random strings. So I changed `os.urandom` to the seedable `random.randint` so that tests can be run deterministically (usage: `python test_server.py & python test_client.py 42` where 42 is the seed)

Turns out it's all good with the random strings, but the problem was 

```
peer.send(0, enet.Packet(msg))
host.service(0)
peer.disconnect()
```

Changed `host.service(0)` to `host.service(100)` to avoid the problem. The issue was subtle: python test_server.py wouldn't stop sometimes because it might not receive SHUTDOWN message. This change fixes it.

FYI: this is the second change I did for the existing tests. First one was changing `"SHUTDOWN"` to `b"SHUTDOWN"` (introduced problems to py3k runs)

To my best knowledge, this PR is ready for the codebase and it indeed looks good to me.

aresch commented:
Thanks for all the work addressing my comments!
